src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1106,83 +1104,151 @@ TargetEntryChangesValue(TargetEntry *targetEntry, Var *column, FromExpr *joinTre
 {
-	Task *modifyTask = NULL;
-	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
-
-	modifyTask = CitusMakeNode(Task);
-	modifyTask->jobId = INVALID_JOB_ID;
-	modifyTask->taskId = INVALID_TASK_ID;
-	modifyTask->taskType = MODIFY_TASK;
-	modifyTask->queryString = NULL;
-	modifyTask->anchorShardId = INVALID_SHARD_ID;
-	modifyTask->dependedTaskList = NIL;
-	modifyTask->replicationModel = cacheEntry->replicationModel;
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	List *taskList = NIL;
+	Job *job = NULL;
+	bool requiresMasterEvaluation = false;
+	bool deferredPruning = false;
 
-	if (originalQuery->onConflict != NULL)
+	if (!CanShardPrune(distributedTableId, query))
 	{
-		RangeTblEntry *rangeTableEntry = NULL;
+		/* there is a non-constant in the partition column, cannot prune yet */
+		taskList = NIL;
+		deferredPruning = true;
 
-		/* set the flag */
-		modifyTask->upsertQuery = true;
-
-		/* setting an alias simplifies deparsing of UPSERTs */
-		rangeTableEntry = linitial(originalQuery->rtable);
-		if (rangeTableEntry->alias == NULL)
+		/* must evaluate the non-constant in the partition column */
+		requiresMasterEvaluation = true;
+	}
+	else
+	{
+		taskList = RouterModifyTaskList(query, planningError);
+		if (*planningError)
 		{
-			Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
-			rangeTableEntry->alias = alias;
+			return NULL;
 		}
+
+		/* determine whether there are function calls to evaluate */
+		requiresMasterEvaluation = RequiresMasterEvaluation(originalQuery);
 	}
 
-	if (shardInterval != NULL)
+	if (!requiresMasterEvaluation)
 	{
-		uint64 shardId = shardInterval->shardId;
-		StringInfo queryString = makeStringInfo();
+		/* no functions or parameters, build the query strings upfront */
+		RebuildQueryStrings(originalQuery, taskList);
+	}
+
+	job = CreateJob(originalQuery);
+	job->taskList = taskList;
+	job->requiresMasterEvaluation = requiresMasterEvaluation;
+	job->deferredPruning = deferredPruning;
+
+	return job;
+}
+
+
+/*
+ * CreateJob returns a new Job for the given query.
+ */
+static Job *
+CreateJob(Query *query)
+{
+	Job *job = NULL;
+
+	job = CitusMakeNode(Job);
+	job->dependedJobList = NIL;
+	job->dependedJobList = NIL;
+	job->jobId = INVALID_JOB_ID;
+	job->subqueryPushdown = false;
+	job->jobQuery = query;
+	job->taskList = NIL;
+	job->requiresMasterEvaluation = false;
+
+	return job;
+}
+
+
+/*
+ * CanShardPrune determines whether a query is ready for shard pruning
+ * by checking whether there is a constant value in the partition column.
+ */
+static bool
+CanShardPrune(Oid distributedTableId, Query *query)
+{
+	uint32 rangeTableId = 1;
+	Var *partitionColumn = NULL;
+	Expr *partitionValueExpr = NULL;
 
-		/* grab shared metadata lock to stop concurrent placement additions */
-		LockShardDistributionMetadata(shardId, ShareLock);
+	if (query->commandType != CMD_INSERT)
+	{
+		/* we assume UPDATE/DELETE is always prunable */
```

Comment:
Mainly that we currently always perform shard pruning for UPDATE/DELETE in the planner. 

If there is no partition column filter or a non-constant filter (e.g. function), then shard pruning simply returns all shards, and this could in the future become a multi-shard UPDATE/DELETE. For INSERT things are a bit different since we really need to prune down to a single shard.

We could at some point decide to defer pruning for UPDATE/DELETE if there is a parameter or function in the partition column filter. In that case we'd change this function.